File: credit.c
#include <cs50.h>
#include <stdio.h>

long SIXTEEN_DIGIT_NUMBER = 1000000000000000;
long FIFTEEN_DIGIT_NUMBER = 100000000000000;
long THIRTEEN_DIGIT_NUMBER = 1000000000000;

string VISA = "VISA";
string AMEX = "AMEX";
string MASTERCARD = "MASTERCARD";
string INVALID = "INVALID";

bool determineIsStartingDigitForMultiplication(long divider);

string determineCardType(long number);

int calculateChecksum(long number);

long getDivider(long number);

int main(void)
{
    long number = 0;

    number = get_long("Type your credit card number: ");

    // get card type according to number length
    // and starting digits
    string cardType = determineCardType(number);

    if (cardType == INVALID)
    {
        printf("%s\n", cardType);

        return 0;
    }

    int checksum = calculateChecksum(number);
    bool isLastDigitZero = checksum % 10 == 0;

    // print the adequate card type
    // if the last digit is zero
    // otherwise print invalid
    if (isLastDigitZero)
    {
        printf("%s\n", cardType);

        return 0;
    }

    printf("%s\n", INVALID);
}

// determine if first digit is the one that will be multiplied
bool determineIsStartingDigitForMultiplication(long divider)
{
    if (divider < FIFTEEN_DIGIT_NUMBER)
    {
        return false;
    }

    else if (divider < SIXTEEN_DIGIT_NUMBER)
    {
        return false;
    }

    else
    {
        return true;
    }
}

int getStartingDigits(long number);

// determine card type based on the
// length of the number and its starting digits
string determineCardType(long number)
{
    string cardType = INVALID;
    int startingDigits = getStartingDigits(number);

    if (number < THIRTEEN_DIGIT_NUMBER)
    {

    }
    else if (number < FIFTEEN_DIGIT_NUMBER)
    {
        if (startingDigits == 4)
        {
            cardType = VISA;
        }
    }

    else if (number < SIXTEEN_DIGIT_NUMBER)
    {
        if (startingDigits == 34 || startingDigits == 37)
        {
            cardType = AMEX;
        }
    }

    else
    {
        if (
            startingDigits == 51
            || startingDigits == 52
            || startingDigits == 53
            || startingDigits == 54
            || startingDigits == 55
        )
        {
            cardType = MASTERCARD;
        }
        else if (startingDigits / 10 == 4)
        {
            cardType = VISA;
        }
    }

    return cardType;
}

// get the starting digits of a number
// based on its length
int getStartingDigits(long number)
{
    if (number < FIFTEEN_DIGIT_NUMBER)
    {
        return number / THIRTEEN_DIGIT_NUMBER;
    }

    else if (number < SIXTEEN_DIGIT_NUMBER)
    {
        return number / (FIFTEEN_DIGIT_NUMBER / 10);
    }

    else
    {
        return number / (SIXTEEN_DIGIT_NUMBER / 10);
    }
}

int calculateChecksum(long number)
{
    // get the largest divider according
    // the length of the number
    long divider = getDivider(number);
    int digit = 0;
    // check if the first digit will be
    // the one that will be multiplied and
    // accordingly multiply every other digit
    // base on the first multiplied digit
    bool isDigitForMultiplication = determineIsStartingDigitForMultiplication(divider);
    int checksum = 0;

    // get each digit and either multiply it
    // and add its digits to the checksum
    // or simpy add the digit to the checksum
    while (divider >= 1)
    {
        digit = (number / divider) % 10;

        if (isDigitForMultiplication)
        {
            digit = digit * 2;

            if (digit >= 10)
            {
                int firstDigit = digit / 10;
                int secondDigit = digit % 10;

                digit = firstDigit + secondDigit;
            }

            checksum += digit;
        }
        else
        {
            checksum += digit;
        }

        isDigitForMultiplication = !isDigitForMultiplication;

        divider /= 10;
    }

    return checksum;
}

// determine the divider based on the
// length of the number
long getDivider(long number)
{
    long result = 0;

    if (number < FIFTEEN_DIGIT_NUMBER)
    {
        return THIRTEEN_DIGIT_NUMBER;
    }

    else if (number < SIXTEEN_DIGIT_NUMBER)
    {
        return FIFTEEN_DIGIT_NUMBER;
    }

    else
    {
        return SIXTEEN_DIGIT_NUMBER;
    }
}

File: credit.py
import re

VISA_REGEX = "^4\d{12,15}$"
AMEX_REGEX = "^(34|37)\d{13}$"
MASTERCARD_REGEX = "^(51|52|53|54|55)\d{14}$"


def main():
    # Getthe input from the user as a string
    card_number: str = str(input("Number: "))

    # Determine the type of card
    card_type: str = determine_card_type(card_number)

    if card_type == "INVALID":
        print(card_type)

        return

    # Calculate the checksum using the Luhn's Algorithm
    checksum: int = calculate_checksum(card_number)

    # Check if last digit is a zero
    is_last_digit_zero: bool = checksum % 10 == 0

    # Print the card type
    if is_last_digit_zero:
        print(card_type)

        return

    print("INVALID")


# Determine the card type based on card regexes
def determine_card_type(card_number: str) -> str:
    card_type: str = "INVALID"

    if re.search(VISA_REGEX, card_number):
        card_type = "VISA"
    elif re.search(AMEX_REGEX, card_number):
        card_type = "AMEX"
    elif re.search(MASTERCARD_REGEX, card_number):
        card_type = "MASTERCARD"

    return card_type


# Calculate checksum by using the Luhn's Algorithm
def calculate_checksum(card_number: str) -> int:
    checksum: int = 0
    should_multiply: bool = False

    # Iterate over the card number backwards and add them to
    # the checksum or add the digits that resulted from multiplicating
    # the digit by 2
    for i in range(len(card_number) - 1, -1, -1):
        if should_multiply:
            number = 2 * int(card_number[i])

            if number >= 10:
                first_digit = number // 10
                second_digit = number % 10
                number = first_digit + second_digit

            checksum += number
        else:
            checksum += int(card_number[i])

        should_multiply = not should_multiply

    return checksum


main()
